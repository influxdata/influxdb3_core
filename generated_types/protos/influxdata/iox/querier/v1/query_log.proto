syntax = "proto3";
package influxdata.iox.querier.v1;
option go_package = "github.com/influxdata/iox/querier/v1";

import "google/protobuf/duration.proto";
import "google/protobuf/timestamp.proto";

import "influxdata/iox/querier/v1/params.proto";

service QueryLogService {
    // Get the process-local query log.
    rpc GetLog(GetLogRequest) returns (stream GetLogResponse);
}

message GetLogRequest {}

// Describe a single query.
message LogEntry {
  // Unique entry ID.
  string id = 1;

  // Current lifecycle phase of the query.
  Phase phase = 16;

  // Namespace ID.
  int64 namespace_id = 2;

  // Namespace name.
  string namespace_name = 3;

  // Query type, e.g. `sql`.
  string query_type = 4;

  // Query text.
  string query_text = 5;

  // Query parameters
  repeated QueryParam query_params = 21;

  // Trace ID.
  string trace_id = 6;

  // Start timestamp.
  google.protobuf.Timestamp issue_time = 7;

  // Number of partitions the query was executed on.
  int64 partitions = 18;

  // Number of parquet files the query was executed on.
  int64 parquet_files = 19;

  reserved 8;
  reserved 'query_completed_duration';

  // Duration it took to acquire a semaphore permit, relative to `issue_time`.
  google.protobuf.Duration permit_duration = 10;

  // Duration it took to plan the query, relative to `issue_time` + `permit_duration`.
  google.protobuf.Duration plan_duration = 11;

  // Duration it took to execute the query, relative to `issue_time` +
  // `permit_duration` + `plan_duration`.
  google.protobuf.Duration execute_duration = 12;

  // Duration from `issue_time` til the query ended somehow.
  google.protobuf.Duration end2end_duration = 13;

  // CPU duration spend for computation.
  google.protobuf.Duration compute_duration = 15;

  // Maximum memory allocated when executing the query.
  int64 max_memory = 20;

  // If the query completed successfully.
  bool success = 9;

  // If the query is currently running (in any state).
  bool running = 14;

  // If the query was cancelled.
  bool cancelled = 17;
}

// Metadata for the entire response.
message Metadata {
  // Maximum size of the query log.
  uint64 max_size = 1;

  // Number of evicted entries due to the "max size" constraint.
  uint64 evicted = 2;
}

message GetLogResponse {
  oneof data {
    Metadata metadata = 1;
    LogEntry entry = 2;
  }
}

enum Phase {
  PHASE_UNSPECIFIED = 0;
  PHASE_RECEIVED = 1;
  PHASE_PLANNED = 2;
  PHASE_PERMIT = 3;
  PHASE_CANCEL = 4;
  PHASE_SUCCESS = 5;
  PHASE_FAIL = 6;
}
