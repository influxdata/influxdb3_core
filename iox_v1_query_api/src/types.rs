use arrow::datatypes::SchemaRef;
use datafusion::physical_plan::SendableRecordBatchStream;
use serde::{Deserialize, Serialize};

use iox_query::query_log::PermitAndToken;

/// UNIX epoch precision.
/// Doc: <https://docs.influxdata.com/influxdb/v1/query_language/spec/#duration-units>
#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq)]
pub enum Precision {
    #[serde(rename = "ns")]
    Nanoseconds,
    #[serde(rename = "u", alias = "µ")]
    Microseconds,
    #[serde(rename = "ms")]
    Milliseconds,
    #[serde(rename = "s")]
    Seconds,
    #[serde(rename = "m")]
    Minutes,
    #[serde(rename = "h")]
    Hours,
    #[serde(rename = "d")]
    Days,
    #[serde(rename = "w")]
    Weeks,
}

/// An executing InfluxQL statement that produces results that can be
/// streamed as CSV.
pub(crate) struct Statement {
    pub schema: SchemaRef,
    pub permit_state: PermitAndToken,
    pub stream: SendableRecordBatchStream,
}

impl Statement {
    pub(crate) fn new(
        schema: SchemaRef,
        permit_state: PermitAndToken,
        stream: SendableRecordBatchStream,
    ) -> Self {
        Self {
            schema,
            permit_state,
            stream,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use serde_json::{from_str, to_string};

    #[test]
    fn test_precision_deserialize() {
        // Test standard rename attributes
        assert_eq!(
            from_str::<Precision>(r#""ns""#).unwrap(),
            Precision::Nanoseconds
        );
        assert_eq!(
            from_str::<Precision>(r#""u""#).unwrap(),
            Precision::Microseconds
        );
        assert_eq!(
            from_str::<Precision>(r#""ms""#).unwrap(),
            Precision::Milliseconds
        );
        assert_eq!(from_str::<Precision>(r#""s""#).unwrap(), Precision::Seconds);
        assert_eq!(from_str::<Precision>(r#""m""#).unwrap(), Precision::Minutes);
        assert_eq!(from_str::<Precision>(r#""h""#).unwrap(), Precision::Hours);
        assert_eq!(from_str::<Precision>(r#""d""#).unwrap(), Precision::Days);
        assert_eq!(from_str::<Precision>(r#""w""#).unwrap(), Precision::Weeks);

        // Test the alias for Microseconds
        assert_eq!(
            from_str::<Precision>(r#""µ""#).unwrap(),
            Precision::Microseconds
        );
    }

    #[test]
    fn test_precision_serialize() {
        // Test that each enum variant serializes to the expected string
        assert_eq!(to_string(&Precision::Nanoseconds).unwrap(), r#""ns""#);
        assert_eq!(to_string(&Precision::Microseconds).unwrap(), r#""u""#); // Note: serializes to "u", not "µ"
        assert_eq!(to_string(&Precision::Milliseconds).unwrap(), r#""ms""#);
        assert_eq!(to_string(&Precision::Seconds).unwrap(), r#""s""#);
        assert_eq!(to_string(&Precision::Minutes).unwrap(), r#""m""#);
        assert_eq!(to_string(&Precision::Hours).unwrap(), r#""h""#);
        assert_eq!(to_string(&Precision::Days).unwrap(), r#""d""#);
        assert_eq!(to_string(&Precision::Weeks).unwrap(), r#""w""#);
    }

    #[test]
    fn test_precision_error_cases() {
        // Test invalid inputs
        let invalid_result = from_str::<Precision>(r#""invalid""#);
        assert!(invalid_result.is_err());

        // Test case sensitivity (serde is case-sensitive by default)
        let uppercase_result = from_str::<Precision>(r#""NS""#);
        assert!(uppercase_result.is_err());
    }

    #[test]
    fn test_precision_roundtrip() {
        // Test serialization and deserialization roundtrip for all variants
        let variants = vec![
            Precision::Nanoseconds,
            Precision::Microseconds,
            Precision::Milliseconds,
            Precision::Seconds,
            Precision::Minutes,
            Precision::Hours,
            Precision::Days,
            Precision::Weeks,
        ];

        for variant in variants {
            let serialized = to_string(&variant).unwrap();
            let deserialized = from_str::<Precision>(&serialized).unwrap();
            assert_eq!(variant, deserialized);
        }
    }
}
