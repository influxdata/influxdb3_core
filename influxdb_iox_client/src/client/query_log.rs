use ::generated_types::google::{FieldViolation, OptionalField};
use client_util::connection::GrpcConnection;
use futures_util::{stream::BoxStream, StreamExt, TryStreamExt};

use self::generated_types::{query_log_service_client::QueryLogServiceClient, *};
use crate::connection::Connection;
use crate::error::Error;

/// Re-export generated_types
pub mod generated_types {
    pub use generated_types::influxdata::iox::querier::v1::*;
}

/// A basic client for working with the query log.
#[derive(Debug, Clone)]
pub struct Client {
    inner: QueryLogServiceClient<GrpcConnection>,
}

impl Client {
    /// Creates a new client with the provided connection
    pub fn new(connection: Connection) -> Self {
        Self {
            inner: QueryLogServiceClient::new(connection.into_grpc_connection()),
        }
    }

    /// Get log.
    pub async fn get_log(&mut self) -> Result<Log, Error> {
        let mut stream = self.inner.get_log(GetLogRequest {}).await?.into_inner();

        // first message should be metadata
        let Some(first_msg) = stream.try_next().await? else {
            return Err(Error::InvalidResponse(FieldViolation::required(
                "first message",
            )));
        };
        let get_log_response::Data::Metadata(metadata) = first_msg.data.unwrap_field("data")?
        else {
            return Err(Error::InvalidResponse(FieldViolation::required(
                "metadata in first message",
            )));
        };

        Ok(Log {
            metadata,
            entries: stream
                .map_err(Error::from)
                .and_then(|msg| async move {
                    let get_log_response::Data::Entry(entry) = msg.data.unwrap_field("data")?
                    else {
                        return Err(Error::InvalidResponse(FieldViolation::required(
                            "entry in tail messages",
                        )));
                    };
                    Ok(entry)
                })
                .boxed(),
        })
    }
}

/// [`get_log`](Client::get_log) response.
pub struct Log {
    /// Metadata.
    pub metadata: Metadata,

    /// Stream of entries.
    pub entries: BoxStream<'static, Result<LogEntry, Error>>,
}

impl std::fmt::Debug for Log {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Log")
            .field("metadata", &self.metadata)
            .field("entries", &"<STREAM>")
            .finish()
    }
}
