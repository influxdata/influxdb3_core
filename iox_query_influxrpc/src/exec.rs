pub(crate) mod fields_pivot_exec;
mod fields_pivot_stream;
mod series_pivot_exec;
mod series_pivot_stream;

pub(crate) use fields_pivot_exec::FieldsPivotExec;
pub(crate) use series_pivot_exec::SeriesPivotExec;
