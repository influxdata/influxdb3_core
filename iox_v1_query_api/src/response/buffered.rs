//! Streams for producing responses where chunking is not enabled.

use super::{Response, SeriesChunk, SeriesChunkMergeStream, SeriesChunkStream, StatementResult};
use crate::{Result, types::Statement};
use datafusion::physical_plan::SendableRecordBatchStream;
use futures::{Stream, ready};
use iox_query::query_log::PermitAndToken;
use std::{
    future::Future,
    pin::Pin,
    task::{Context, Poll},
};

/// A stream of one [Response] value where the response contains the
/// result data for all the corresponding statements.
pub(crate) struct BufferedResponseStream {
    statements: Vec<Pin<Box<dyn Future<Output = Result<Statement>> + Send>>>,
    statement_id: usize,
    current_statement: Option<(
        PermitAndToken,
        BufferedResultStream<SeriesChunkStream<SendableRecordBatchStream>>,
    )>,
    // Buffer the statment results for all the statements in one response
    response: Option<Response>,
}

impl BufferedResponseStream {
    pub(crate) fn new(statements: Vec<Box<dyn Future<Output = Result<Statement>> + Send>>) -> Self {
        let response = (!statements.is_empty()).then_some(Response::default());

        let statements = statements.into_iter().map(Box::into_pin).collect();
        Self {
            statements,
            statement_id: 0,
            current_statement: None,
            response,
        }
    }

    fn poll_statement(&mut self, cx: &mut Context<'_>) -> Poll<Result<Statement>> {
        self.statements[self.statement_id].as_mut().poll(cx)
    }

    fn poll_next_result(&mut self, cx: &mut Context<'_>) -> Poll<Option<StatementResult>> {
        let (_, stream) = self
            .current_statement
            .as_mut()
            .expect("current_statement is None");
        Pin::new(stream).poll_next(cx)
    }

    fn poll_next_unpin(&mut self, cx: &mut Context<'_>) -> Poll<Option<Response>> {
        if self.statement_id >= self.statements.len() {
            return Poll::Ready(self.response.take());
        }

        if self.current_statement.is_some() {
            match ready!(self.poll_next_result(cx)) {
                Some(result) => {
                    if result.is_error() {
                        let (permit_state, _) = self.current_statement.take().unwrap(); // safe to unwrap because we just checked above
                        permit_state.query_completed_token.fail();
                        self.statement_id += 1;
                    }
                    self.response.as_mut().unwrap().add_result(result); // safe to unwrap because we just checked above
                }
                None => {
                    let (permit_state, _) = self.current_statement.take().unwrap(); // safe to unwrap because we just checked above
                    permit_state.query_completed_token.success();
                    self.statement_id += 1;
                }
            }
        } else {
            match ready!(self.poll_statement(cx)) {
                Ok(Statement {
                    schema,
                    permit_state,
                    stream,
                }) => match SeriesChunkStream::try_new(stream, schema) {
                    Ok(stream) => {
                        self.current_statement = Some((
                            permit_state,
                            BufferedResultStream::new(stream, self.statement_id),
                        ));
                    }
                    Err(e) => {
                        permit_state.query_completed_token.fail();
                        let mut result = StatementResult::new(self.statement_id);
                        result.set_error(e.to_string());
                        self.statement_id += 1;
                        self.response.as_mut().unwrap().add_result(result); // safe to unwrap because we just checked above
                    }
                },
                Err(e) => {
                    let mut result = StatementResult::new(self.statement_id);
                    result.set_error(e.to_string());
                    self.statement_id += 1;
                    self.response.as_mut().unwrap().add_result(result); // safe to unwrap because we just checked above
                }
            }
        }
        self.poll_next_unpin(cx)
    }
}

impl Stream for BufferedResponseStream {
    type Item = Response;

    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        let this = self.get_mut();
        this.poll_next_unpin(cx)
    }
}

pub(super) struct BufferedResultStream<S> {
    inner: SeriesChunkMergeStream<S>,
    result: Option<StatementResult>,
}

impl<S> BufferedResultStream<S> {
    pub(super) fn new(stream: S, statement_id: usize) -> Self {
        let inner = SeriesChunkMergeStream::new(stream, None);
        Self {
            inner,
            result: Some(StatementResult::new(statement_id)),
        }
    }
}

impl<S> BufferedResultStream<S>
where
    S: Stream<Item = Result<SeriesChunk>> + Unpin,
{
    fn poll_next_unpin(&mut self, cx: &mut Context<'_>) -> Poll<Option<StatementResult>> {
        loop {
            if self.result.is_none() {
                return Poll::Ready(None);
            }

            match ready!(Pin::new(&mut self.inner).poll_next(cx)) {
                Some(Ok(chunk)) => {
                    self.result.as_mut().unwrap().add_series(chunk); // safe to unwrap because we just checked above
                }
                Some(Err(e)) => {
                    let mut result = self.result.take().unwrap(); // safe to unwrap because we just checked above
                    result.set_error(e.to_string());
                    return Poll::Ready(Some(result));
                }
                None => return Poll::Ready(self.result.take()),
            }
        }
    }
}

impl<S> Stream for BufferedResultStream<S>
where
    S: Stream<Item = Result<SeriesChunk>> + Unpin,
{
    type Item = StatementResult;

    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        let this = self.get_mut();
        this.poll_next_unpin(cx)
    }
}

#[cfg(test)]
mod tests {
    use crate::response::ResponseSerializer;
    use crate::types::Precision;

    use super::super::tests::{Column, make_statement};
    use super::*;
    use arrow::array::{DictionaryArray, Float64Array, StringArray, TimestampNanosecondArray};
    use arrow::datatypes::Int32Type;
    use futures::StreamExt;
    use iox_query::QueryDatabase;
    use iox_query::exec::IOxSessionContext;
    use iox_query::query_log::QueryLog;
    use iox_query::test::TestDatabaseStore;
    use iox_time::SystemProvider;
    use serde::ser::Serialize;
    use serde_json::ser::Serializer;
    use std::sync::Arc;

    #[tokio::test]
    async fn no_statements() {
        let mut stream = BufferedResponseStream::new(vec![]);
        assert!(stream.next().await.is_none());
    }

    #[tokio::test]
    async fn single_statement_single_series() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let statement = make_statement(
            &db,
            &ctx,
            &log,
            [
                Column::Measurement,
                Column::Time,
                Column::Tag {
                    name: "t1",
                    group_by: false,
                    projected: false,
                },
                Column::Field { name: "f1" },
            ],
            vec![
                Arc::new(StringArray::from(vec!["m1", "m1"])),
                Arc::new(TimestampNanosecondArray::from(vec![1000000000, 2000000000])),
                Arc::new(DictionaryArray::<Int32Type>::from_iter(vec!["t1a", "t1a"])),
                Arc::new(Float64Array::from(vec![1.0, 2.0])),
            ],
        );
        insta::assert_snapshot!(collect_output(BufferedResponseStream::new(vec![statement])).await);
    }

    #[tokio::test]
    async fn single_statement_multi_series() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let statement = make_statement(
            &db,
            &ctx,
            &log,
            [
                Column::Measurement,
                Column::Time,
                Column::Tag {
                    name: "t1",
                    group_by: true,
                    projected: false,
                },
                Column::Field { name: "f1" },
            ],
            vec![
                Arc::new(StringArray::from(vec!["m1", "m1", "m1"])),
                Arc::new(TimestampNanosecondArray::from(vec![
                    1000000000, 2000000000, 3000000000,
                ])),
                Arc::new(DictionaryArray::<Int32Type>::from_iter(vec![
                    "t1a", "t1a", "t1b",
                ])),
                Arc::new(Float64Array::from(vec![1.0, 2.0, 3.0])),
            ],
        );

        insta::assert_snapshot!(collect_output(BufferedResponseStream::new(vec![statement])).await);
    }

    #[tokio::test]
    async fn many_statements() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let statement1 = make_statement(
            &db,
            &ctx,
            &log,
            [
                Column::Measurement,
                Column::Time,
                Column::Tag {
                    name: "t1",
                    group_by: false,
                    projected: false,
                },
                Column::Field { name: "f1" },
            ],
            vec![
                Arc::new(StringArray::from(vec!["m1", "m1", "m1", "m2"])),
                Arc::new(TimestampNanosecondArray::from(vec![
                    1000000000, 2000000000, 3000000000, 1000000000,
                ])),
                Arc::new(DictionaryArray::<Int32Type>::from_iter(vec![
                    "t1a", "t1a", "t1a", "t1a",
                ])),
                Arc::new(Float64Array::from(vec![1.0, 2.0, 3.0, 1.0])),
            ],
        );
        let statement2 = make_statement(
            &db,
            &ctx,
            &log,
            [
                Column::Measurement,
                Column::Time,
                Column::Tag {
                    name: "t1",
                    group_by: true,
                    projected: false,
                },
                Column::Field { name: "f1" },
            ],
            vec![
                Arc::new(StringArray::from(vec!["m1", "m1", "m1", "m2"])),
                Arc::new(TimestampNanosecondArray::from(vec![
                    1000000000, 2000000000, 3000000000, 1000000000,
                ])),
                Arc::new(DictionaryArray::<Int32Type>::from_iter(vec![
                    "t1a", "t1a", "t1b", "t1b",
                ])),
                Arc::new(Float64Array::from(vec![1.0, 2.0, 3.0, 1.0])),
            ],
        );

        insta::assert_snapshot!(
            collect_output(BufferedResponseStream::new(vec![statement1, statement2])).await
        );
    }

    #[tokio::test]
    async fn test_epoch() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let statement = make_statement(
            &db,
            &ctx,
            &log,
            [
                Column::Measurement,
                Column::Time,
                Column::Tag {
                    name: "t1",
                    group_by: true,
                    projected: false,
                },
                Column::Field { name: "f1" },
            ],
            vec![
                Arc::new(StringArray::from(vec!["m1", "m1", "m1"])),
                Arc::new(TimestampNanosecondArray::from(vec![
                    1000000000, 2000000000, 3000000000,
                ])),
                Arc::new(DictionaryArray::<Int32Type>::from_iter(vec![
                    "t1a", "t1a", "t1b",
                ])),
                Arc::new(Float64Array::from(vec![1.0, 2.0, 3.0])),
            ],
        );

        insta::assert_snapshot!(
            collect_output_epoch(
                BufferedResponseStream::new(vec![statement]),
                Some(Precision::Milliseconds)
            )
            .await
        );
    }

    async fn collect_output(stream: impl Stream<Item = Response> + Send) -> String {
        collect_output_epoch(stream, None).await
    }

    async fn collect_output_epoch(
        stream: impl Stream<Item = Response> + Send,
        epoch: Option<Precision>,
    ) -> String {
        stream
            .map(|r| {
                let mut w = vec![];
                let mut ser = Serializer::new(&mut w);
                let r = ResponseSerializer::new(&r, epoch, true);
                r.serialize(&mut ser).unwrap();
                w.push(b'\n');
                String::from_utf8(w).unwrap()
            })
            .collect()
            .await
    }
}
