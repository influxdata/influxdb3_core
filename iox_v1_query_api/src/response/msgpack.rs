//! Message Pack encoding of InfluxQL query results.
use crate::types::Precision;

use super::{BufferedResponseStream, ChunkedResponseStream, Response, ResponseSerializer};
use bytes::buf::BufMut;
use bytes::{Bytes, BytesMut};
use futures::Stream;
use rmp_serde::Serializer;
use serde::Serialize;
use std::{
    pin::Pin,
    task::{Context, Poll},
};
use tracing::warn;

/// A generic Message Pack-encoded [Response] stream.
pub(crate) struct MessagePackStream<S> {
    stream: S,
    epoch: Option<Precision>,
}

impl<S> MessagePackStream<S> {
    pub(crate) fn new(stream: S, epoch: Option<Precision>) -> Self {
        Self { stream, epoch }
    }

    fn poll_next_inner(&mut self, cx: &mut Context<'_>) -> Poll<Option<Response>>
    where
        S: Stream<Item = Response> + Unpin,
    {
        Pin::new(&mut self.stream).poll_next(cx)
    }
}

impl<S> Stream for MessagePackStream<S>
where
    S: Stream<Item = Response> + Unpin,
{
    type Item = Bytes;

    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        let this = self.get_mut();
        this.poll_next_inner(cx).map(|opt| {
            opt.and_then(|resp| {
                let mut w = BytesMut::new().writer();
                let mut serializer = Serializer::new(&mut w);
                let resp = ResponseSerializer::new(&resp, this.epoch, true);
                if let Err(e) = resp.serialize(&mut serializer) {
                    warn!(error = %e, "failed to serialize response");
                    None
                } else {
                    Some(w.into_inner().freeze())
                }
            })
        })
    }
}

pub(crate) type ChunkedMessagePackStream = MessagePackStream<ChunkedResponseStream>;
pub(crate) type BufferedMessagePackStream = MessagePackStream<BufferedResponseStream>;

#[cfg(test)]
mod tests {
    use crate::response::buffered::BufferedResponseStream;

    use super::super::tests::{Column, make_statement};
    use super::*;
    use arrow::array::{
        ArrayRef, DictionaryArray, Float64Array, StringArray, TimestampNanosecondArray,
    };
    use arrow::datatypes::Int32Type;
    use futures::StreamExt;
    use iox_query::QueryDatabase;
    use iox_query::exec::IOxSessionContext;
    use iox_query::query_log::QueryLog;
    use iox_query::test::TestDatabaseStore;
    use iox_time::SystemProvider;
    use rmp_serde::Deserializer;
    use serde::Deserialize;
    use serde_json::Value;
    use std::sync::Arc;

    #[tokio::test]
    async fn empty_stream() {
        let stream = ChunkedResponseStream::new(vec![], 2);
        let mut stream = ChunkedMessagePackStream::new(stream, None);
        assert!(stream.next().await.is_none());

        let stream = BufferedResponseStream::new(vec![]);
        let mut stream = BufferedMessagePackStream::new(stream, None);
        assert!(stream.next().await.is_none());
    }

    #[tokio::test]
    async fn single_chunk() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let columns = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: false,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1"])),
            Arc::new(TimestampNanosecondArray::from(vec![1000000000, 2000000000])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec!["t1a", "t1a"])),
            Arc::new(Float64Array::from(vec![1.0, 2.0])),
        ];

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = ChunkedResponseStream::new(vec![statement], 2);
        let stream = ChunkedMessagePackStream::new(stream, None);
        insta::assert_snapshot!(collect_output(stream).await);

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = BufferedResponseStream::new(vec![statement]);
        let stream = BufferedMessagePackStream::new(stream, None);
        insta::assert_snapshot!(collect_output(stream).await);
    }

    #[tokio::test]
    async fn many_chunks() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let columns = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: false,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1", "m1"])),
            Arc::new(TimestampNanosecondArray::from(vec![
                1000000000, 2000000000, 3000000000,
            ])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec![
                "t1a", "t1a", "t1a",
            ])),
            Arc::new(Float64Array::from(vec![1.0, 2.0, 3.0])),
        ];

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = ChunkedResponseStream::new(vec![statement], 2);
        let stream = ChunkedMessagePackStream::new(stream, None);
        insta::assert_snapshot!(collect_output(stream).await);

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = BufferedResponseStream::new(vec![statement]);
        let stream = BufferedMessagePackStream::new(stream, None);
        insta::assert_snapshot!(collect_output(stream).await);
    }

    #[tokio::test]
    async fn many_chunks_many_measurments() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let columns = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: false,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1", "m1", "m2"])),
            Arc::new(TimestampNanosecondArray::from(vec![
                1000000000, 2000000000, 3000000000, 1000000000,
            ])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec![
                "t1a", "t1a", "t1a", "t1a",
            ])),
            Arc::new(Float64Array::from(vec![1.0, 2.0, 3.0, 1.0])),
        ];

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = ChunkedResponseStream::new(vec![statement], 2);
        let stream = ChunkedMessagePackStream::new(stream, None);
        insta::assert_snapshot!(collect_output(stream).await);

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = BufferedResponseStream::new(vec![statement]);
        let stream = BufferedMessagePackStream::new(stream, None);
        insta::assert_snapshot!(collect_output(stream).await);
    }

    #[tokio::test]
    async fn many_statements() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let columns1 = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: false,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data1: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1", "m1", "m2"])),
            Arc::new(TimestampNanosecondArray::from(vec![
                1000000000, 2000000000, 3000000000, 1000000000,
            ])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec![
                "t1a", "t1a", "t1a", "t1a",
            ])),
            Arc::new(Float64Array::from(vec![1.0, 2.0, 3.0, 1.0])),
        ];

        let columns2 = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: true,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data2: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1", "m1", "m2"])),
            Arc::new(TimestampNanosecondArray::from(vec![
                1000000000, 2000000000, 3000000000, 1000000000,
            ])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec![
                "t1a", "t1a", "t1b", "t1b",
            ])),
            Arc::new(Float64Array::from(vec![1.0, 2.0, 3.0, 1.0])),
        ];

        let statement1 = make_statement(&db, &ctx, &log, columns1.clone(), data1.clone());
        let statement2 = make_statement(&db, &ctx, &log, columns2.clone(), data2.clone());
        let stream = ChunkedResponseStream::new(vec![statement1, statement2], 2);
        let stream = ChunkedMessagePackStream::new(stream, None);
        insta::assert_snapshot!(collect_output(stream).await);

        let statement1 = make_statement(&db, &ctx, &log, columns1.clone(), data1.clone());
        let statement2 = make_statement(&db, &ctx, &log, columns2.clone(), data2.clone());
        let stream = BufferedResponseStream::new(vec![statement1, statement2]);
        let stream = BufferedMessagePackStream::new(stream, None);
        insta::assert_snapshot!(collect_output(stream).await);
    }

    #[tokio::test]
    async fn test_epoch() {
        let db: Arc<dyn QueryDatabase> = Arc::new(TestDatabaseStore::default());
        let ctx = Arc::new(IOxSessionContext::with_testing());
        let log = Arc::new(QueryLog::new(
            1,
            Arc::new(SystemProvider::new()),
            &metric::Registry::new(),
            None,
        ));
        let columns = [
            Column::Measurement,
            Column::Time,
            Column::Tag {
                name: "t1",
                group_by: false,
                projected: false,
            },
            Column::Field { name: "f1" },
        ];
        let data: Vec<ArrayRef> = vec![
            Arc::new(StringArray::from(vec!["m1", "m1"])),
            Arc::new(TimestampNanosecondArray::from(vec![1000000000, 2000000000])),
            Arc::new(DictionaryArray::<Int32Type>::from_iter(vec!["t1a", "t1a"])),
            Arc::new(Float64Array::from(vec![1.0, 2.0])),
        ];

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = ChunkedResponseStream::new(vec![statement], 2);
        let stream = ChunkedMessagePackStream::new(stream, Some(Precision::Microseconds));
        insta::assert_snapshot!(collect_output(stream).await);

        let statement = make_statement(&db, &ctx, &log, columns.clone(), data.clone());
        let stream = BufferedResponseStream::new(vec![statement]);
        let stream = BufferedMessagePackStream::new(stream, Some(Precision::Microseconds));
        insta::assert_snapshot!(collect_output(stream).await);
    }

    async fn collect_output<S: Stream<Item = Bytes> + Send>(stream: S) -> String {
        String::from_utf8(
            stream
                .map(Vec::<u8>::from)
                .map(|v| {
                    // Docode the msgpack and recode as JSON to make it
                    // easier to validate.
                    let mut de = Deserializer::new(&v[..]);
                    let value = Value::deserialize(&mut de).unwrap();
                    let mut v = serde_json::to_vec(&value).unwrap();
                    v.push(b'\n');
                    v
                })
                .concat()
                .await,
        )
        .unwrap()
    }
}
