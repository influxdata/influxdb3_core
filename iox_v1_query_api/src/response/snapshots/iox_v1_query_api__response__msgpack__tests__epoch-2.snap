---
source: iox_v1_query_api/src/response/msgpack.rs
expression: collect_output(stream).await
---
[[[0,[["m1",["time","t1","f1"],[[1000000,"t1a",1],[2000000,"t1a",2]]]]]]]
